Mathlib/Analysis/Meromorphic/NormalFormAt.lean

```diff
@@ -0,0 +1,240 @@
+
+/-!
+## Continuous extension and conversion to normal form
+-/
+
+variable (f) (x) in
+/-- If `f` is meromorphic at `x`, convert `f` to normal form at `x` by changing its value at `x`.
+Otherwise, returns the 0 function. -/
+noncomputable def toMeromorphicNFAt :
+    𝕜 → E := by
+  by_cases hf : MeromorphicAt f x
+  · classical -- do not complain about decidability issues in Function.update
+    apply Function.update f x
+    by_cases h₁f : hf.order = (0 : ℤ)
+    · rw [hf.order_eq_int_iff] at h₁f
+      exact (Classical.choose h₁f) x
+    · exact 0
+  · exact 0
+
+/-- Conversion to normal form at `x` by changes the value only at x. -/
+lemma MeromorphicAt.eqOn_compl_singleton_toNF (hf : MeromorphicAt f x) :
+    Set.EqOn f (toMeromorphicNFAt f x) {x}ᶜ :=
+  fun _ _ ↦ by simp_all [toMeromorphicNFAt]
+
+/-- Conversion to normal form at `x` changes the value only at x. -/
+lemma MeromorphicAt.toNF_id_on_nhdNE (hf : MeromorphicAt f x) :
+    f =ᶠ[𝓝[≠] x] toMeromorphicNFAt f x :=
+  eventually_nhdsWithin_of_forall (fun _ hz ↦ hf.eqOn_compl_singleton_toNF hz)
+
+-- Two analytic functions agree on a punctured neighborhood iff they agree on a neighborhood.
+private lemma AnalyticAt.eventuallyEq_nhdNE_iff_eventuallyEq_nhd {g : 𝕜 → E} {z₀ : 𝕜}
+  (hf : AnalyticAt 𝕜 f z₀) (hg : AnalyticAt 𝕜 g z₀) (hfg : f =ᶠ[𝓝[≠] z₀] g) :
+    f =ᶠ[𝓝 z₀] g := by
+  rcases ((hf.sub hg).eventually_eq_zero_or_eventually_ne_zero) with h | h
+  · exact Filter.eventuallyEq_iff_sub.2 h
+  · simpa using (Filter.eventually_and.2 ⟨Filter.eventuallyEq_iff_sub.mp hfg, h⟩).exists
+
+/-- After conversion to normal form at `x`, the function has normal form. -/
+theorem MeromorphicAt.meromorphicNFAt_toNF (hf : MeromorphicAt f x) :
+    MeromorphicNFAt (toMeromorphicNFAt f x) x := by
+  by_cases h₂f : hf.order = ⊤
+  · have : toMeromorphicNFAt f x =ᶠ[𝓝 x] 0 := by
+      apply eventuallyEq_nhds_of_eventuallyEq_nhdsNE
+      · exact hf.toNF_id_on_nhdNE.symm.trans (hf.order_eq_top_iff.1 h₂f)
+      · simp [h₂f, toMeromorphicNFAt, hf]
+    apply AnalyticAt.MeromorphicNFAt
+    rw [analyticAt_congr this]
+    exact analyticAt_const
+  · lift hf.order to ℤ using h₂f with n hn
+    obtain ⟨g, h₁g, h₂g, h₃g⟩ := (hf.order_eq_int_iff n).1 hn.symm
+    right
+    use n, g, h₁g, h₂g
+    apply eventuallyEq_nhds_of_eventuallyEq_nhdsNE (hf.toNF_id_on_nhdNE.symm.trans h₃g)
+    simp only [toMeromorphicNFAt, hf, ↓reduceDIte, ← hn, WithTop.coe_zero,
+      WithTop.coe_eq_zero, ne_eq, Function.update_self, sub_self]
+    split_ifs with h₃f
+    · obtain ⟨h₁G, _, h₃G⟩ := Classical.choose_spec ((hf.order_eq_int_iff 0).1 (h₃f ▸ hn.symm))
+      apply Filter.EventuallyEq.eq_of_nhds
+      apply h₁G.eventuallyEq_nhdNE_iff_eventuallyEq_nhd (by fun_prop)
+      filter_upwards [h₃g, h₃G]
+      simp_all
+    · simp [h₃f, zero_zpow]
+
+/-- If `f` has normal form at `x`, then `f` equals `f.toNF`. -/
+theorem MeromorphicNFAt.toNF_eq_id (hf : MeromorphicNFAt f x) :
+    f = toMeromorphicNFAt f x := by
```

Comment:
```suggestion
theorem MeromorphicNFAt.toNF_eq_self (hf : MeromorphicNFAt f x) :
    toMeromorphicNFAt f x = f := by
```
and can you prove the iff version?

Reply:
Will do.

Reply:
Done. Along the way, I remove an assumption from `meromorphicNFAt_toMeromorphicNFAt` and made the name more descriptive.